libkbfs/folder_branch_ops.go

```diff
@@ -5449,31 +5450,54 @@ func (fbo *folderBranchOps) backgroundFlusher(betweenFlushes time.Duration) {
 			// so don't bother waiting for a signal, just get right to
 			// the main attraction.
 			doSelect = false
+		} else if fbo.getCachedDirOpsCount(lState) >=
+			fbo.config.BGFlushDirOpBatchSize() {
+			doSelect = false
 		}
 
 		if doSelect {
 			// Wait until we really have a write waiting.
-			forced := false
+			doWait := true
 			select {
 			case <-fbo.syncNeededChan:
+				if fbo.getCachedDirOpsCount(lState) >=
+					fbo.config.BGFlushDirOpBatchSize() {
+					doWait = false
+				}
 			case <-fbo.forceSyncChan:
-				forced = true
+				doWait = false
 			case <-fbo.shutdownChan:
 				return
 			}
 
-			if !forced {
-				select {
-				case <-ticker.C:
-				case <-fbo.forceSyncChan:
-				case <-fbo.shutdownChan:
-					return
+			if doWait {
+				timer := time.NewTimer(fbo.config.BGFlushPeriod())
+				// Loop until either a tick's worth of time passes,
+				// the batch size of directory ops is full, a sync is
+				// forced, or a shutdown happens.
+			loop:
+				for {
+					select {
+					case <-timer.C:
+						break loop
+					case <-fbo.syncNeededChan:
+						if fbo.getCachedDirOpsCount(lState) >=
+							fbo.config.BGFlushDirOpBatchSize() {
+							break loop
+						}
+					case <-fbo.forceSyncChan:
+						break loop
+					case <-fbo.shutdownChan:
+						return
+					}
 				}
+				timer.Stop()
```

Comment:
Don't need this, I think -- it's only tickers that have resources that need to be freed.